WordPress/Classes/ViewRelated/System/WPSplitViewController.swift

```diff
@@ -0,0 +1,342 @@
+        guard let primaryNavigationController = primaryViewController as? UINavigationController else {
+            assertionFailure("Split view's primary view controller should be a navigation controller!")
+            return nil
+        }
+
+        var viewControllers: [UIViewController] = []
+
+        // Splits the view controller list into primary and detail view controllers at the specified index
+        let separateViewControllersAtIndex: (Int -> Void) = { index in
+            viewControllers = Array(primaryNavigationController.viewControllers.suffixFrom(index))
+            primaryNavigationController.viewControllers = Array(primaryNavigationController.viewControllers.prefixUpTo(index))
+        }
+
+        if let index = primaryNavigationController.viewControllers.indexOf({ $0 is UINavigationController }) {
+            // If there's another navigation controller somewhere in the primary navigation stack
+            // (this is the default behaviour of a collapse), then we'll split the view controllers
+            // apart at that point.
+            separateViewControllersAtIndex(index)
+        } else if let index = primaryNavigationController.viewControllers.lastIndexOf({ $0 is WPSplitViewControllerDetailProvider }) {
+            // Otherwise, if there's a detail provider somewhere in the stack, find the last one
+            separateViewControllersAtIndex(index + 1)
+        }
+
+        dimDetailViewControllerIfNecessary()
+
+        // If we have no detail view controllers, try and fetch the primary view controller's
+        // initial detail view controller.
+        if viewControllers.count == 0 {
+            if let primaryViewController = primaryNavigationController.viewControllers.last,
+                let detailViewController = initialDetailViewControllerForPrimaryViewController(primaryViewController) {
+                return detailViewController
+            }
+        }
+
+        if let firstViewController = viewControllers.first as? UINavigationController {
+            // If it's already a navigation controller, just return it
+            return firstViewController
+        } else {
+            let navigationController = UINavigationController()
+            navigationController.delegate = self
+            navigationController.viewControllers = viewControllers
+
+            return navigationController
+        }
+    }
+
+    func splitViewController(splitViewController: UISplitViewController, collapseSecondaryViewController secondaryViewController: UIViewController, ontoPrimaryViewController primaryViewController: UIViewController) -> Bool {
+        let detailDimmed = dimmingView.superview != nil
+
+        if dimsDetailViewControllerAutomatically {
+            dimDetailViewController(false)
+        }
+
+        // If we've dimmed the detail view controller, then return true so that we
+        // discard it and just show the primary view controller
+        if detailDimmed {
```

Comment:
I _think_ this is basing the logic off of the UI behavior for dimming. Maybe we have another method for detecting whether or not to `return` here that's not tied to UI logic? I may be off on this.